File: packages/py/src/airlock/interface/approver/cli_decision_source.py
"""An interactive terminal approver. I/O entry point — not unit-tested; the
testable path is `auto_approve_decision_source`."""

import json

from airlock.domain.approval import (
    ApprovalDecision,
    ApprovalRequest,
    ApproveDecision,
    RejectDecision,
)
from airlock.interface.approver.approver import DecisionSource


def cli_decision_source(approver: str) -> DecisionSource:
    async def decide(request: ApprovalRequest) -> ApprovalDecision:
        print(f"\n[APPROVAL NEEDED] {request.tool_call.name}({json.dumps(request.tool_call.args)})")
        answer = input("approve / reject? [a/r]: ").strip().lower()
        if answer == "a":
            return ApproveDecision(approver=approver)
        return RejectDecision(approver=approver, reason="Rejected by the operator")

    return decide

File: packages/ts/src/interface/approver/cli-decision-source.ts
import { createInterface } from "node:readline/promises";

import { type ApprovalDecision, DecisionType } from "../../domain/approval";
import type { ApprovalRequest } from "../../domain/approval";
import type { DecisionSource } from "./approver";

/**
 * A decision source that prompts a human on the terminal: it prints the proposed
 * action and reads approve / reject from stdin. This is an I/O entry point and is
 * not unit-tested; the testable path is `autoApproveDecisionSource`.
 */
export function cliDecisionSource(approver: string): DecisionSource {
  return async (request: ApprovalRequest): Promise<ApprovalDecision> => {
    const readline = createInterface({ input: process.stdin, output: process.stdout });
    try {
      const { name, args } = request.toolCall;
      process.stdout.write(`\n[APPROVAL NEEDED] ${name}(${JSON.stringify(args)})\n`);
      const answer = (await readline.question("approve / reject? [a/r]: ")).trim().toLowerCase();
      if (answer === "a") {
        return { type: DecisionType.Approve, approver };
      }
      return { type: DecisionType.Reject, approver, reason: "Rejected by the operator" };
    } finally {
      readline.close();
    }
  };
}
